File: api-java-sql/src/main/java/com/nebulahotel/controller/PagamentosController.java
package com.nebulahotel.controller;

import com.nebulahotel.model.Pagamentos;
import com.nebulahotel.repository.PagamentosRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@CrossOrigin
@RestController
@RequestMapping("/pagamentos")
public class PagamentosController {

    private final PagamentosRepository pagamentosRepository;

    public PagamentosController(PagamentosRepository pagamentosRepository) {
        this.pagamentosRepository = pagamentosRepository;
    }

    @GetMapping
    public List<Pagamentos> listarTodos() {
        return pagamentosRepository.findAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Pagamentos> buscarPorId(@PathVariable Integer id) {
        return pagamentosRepository.findById(id)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @PostMapping
    public ResponseEntity<?> criar(@RequestBody Pagamentos pagamento) {
        if (pagamento.getIdPagamento() == null) {
            return ResponseEntity.badRequest().body(Map.of("erro", "O campo idPagamento é obrigatório."));
        }

        if (pagamentosRepository.existsById(pagamento.getIdPagamento())) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(Map.of("erro", "Já existe pagamento com este idPagamento."));
        }

        Pagamentos novoPagamento = pagamentosRepository.save(pagamento);
        return ResponseEntity.status(HttpStatus.CREATED).body(novoPagamento);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> atualizar(@PathVariable Integer id, @RequestBody Pagamentos pagamento) {
        Optional<Pagamentos> optional = pagamentosRepository.findById(id);
        if (optional.isEmpty()) {
            return ResponseEntity.notFound().build();
        }

        Pagamentos existente = optional.get();
        if (pagamento.getValor() != null) {
            existente.setValor(pagamento.getValor());
        }
        if (pagamento.getMetodo() != null) {
            existente.setMetodo(pagamento.getMetodo());
        }
        if (pagamento.getDataPagamento() != null) {
            existente.setDataPagamento(pagamento.getDataPagamento());
        }
        if (pagamento.getStatus() != null) {
            existente.setStatus(pagamento.getStatus());
        }
        if (pagamento.getTransacaoId() != null) {
            existente.setTransacaoId(pagamento.getTransacaoId());
        }

        Pagamentos atualizado = pagamentosRepository.save(existente);
        return ResponseEntity.ok(atualizado);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deletar(@PathVariable Integer id) {
        if (!pagamentosRepository.existsById(id)) {
            return ResponseEntity.notFound().build();
        }

        pagamentosRepository.deleteById(id);
        return ResponseEntity.noContent().build();
    }
}

File: api-kotlin-nosql/src/main/kotlin/com/nebulahotel/controller/PagamentosController.kt
package com.nebulahotel.controller

import com.nebulahotel.model.Pagamentos
import com.nebulahotel.repository.PagamentosRepository
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@CrossOrigin
@RestController
@RequestMapping("/pagamentos")
class PagamentosController(
    private val pagamentosRepository: PagamentosRepository
) {

    @GetMapping
    fun listarTodos(): List<Pagamentos> = pagamentosRepository.findAll()

    @GetMapping("/{id}")
    fun buscarPorId(@PathVariable id: Int): ResponseEntity<Pagamentos> {
        val pagamento = pagamentosRepository.findById(id)
        return if (pagamento.isPresent) ResponseEntity.ok(pagamento.get()) else ResponseEntity.notFound().build<Pagamentos>()
    }

    @PostMapping
    fun criar(@RequestBody pagamento: Pagamentos): ResponseEntity<Any> {
        if (pagamentosRepository.existsById(pagamento.idPagamento)) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                .body(mapOf("erro" to "Já existe pagamento com este _id."))
        }
        return ResponseEntity.status(HttpStatus.CREATED).body(pagamentosRepository.save(pagamento))
    }

    @PutMapping("/{id}")
    fun atualizar(@PathVariable id: Int, @RequestBody pagamentoAtualizado: Pagamentos): ResponseEntity<Pagamentos> {
        if (!pagamentosRepository.existsById(id)) {
            return ResponseEntity.notFound().build<Pagamentos>()
        }
        val salvo = pagamentosRepository.save(pagamentoAtualizado.copy(idPagamento = id))
        return ResponseEntity.ok(salvo)
    }

    @DeleteMapping("/{id}")
    fun remover(@PathVariable id: Int): ResponseEntity<Void> {
        if (!pagamentosRepository.existsById(id)) {
            return ResponseEntity.notFound().build()
        }
        pagamentosRepository.deleteById(id)
        return ResponseEntity.noContent().build()
    }
}
